myia/parser.py

```diff
@@ -0,0 +1,434 @@
+
+    def process_return(self, block: 'Block', node: ast.Return) -> 'Block':
+        """Process a return statement."""
+        inputs = [self.environment.ast_map[ast.Return],
+                  self.process_expression(block, node.value)]
+        return_ = Apply(inputs, block.graph)
+        block.graph.return_ = return_
+        return_.debug.ast = node
+        return block
+
+    def process_assign(self, block: 'Block', node: ast.Assign) -> 'Block':
+        """Process an assignment."""
+        anf_node = self.process_expression(block, node.value)
+        if len(node.targets) == 1 and isinstance(node.targets[0], ast.Name):
+            anf_node.debug.name = node.targets[0].id
+            block.write(node.targets[0].id, anf_node)
+        else:
+            raise NotImplementedError(node.targets)
+        return block
+
+    def process_expression(self, block: 'Block', node: ast.expr) -> ANFNode:
+        """Process an expression."""
+        expr: ANFNode
+        if isinstance(node, ast.BinOp):
+            inputs_: List[ANFNode] = [self.environment.ast_map[type(node.op)],
+                                      self.process_expression(block,
+                                                              node.left),
+                                      self.process_expression(block,
+                                                              node.right)]
+            expr = Apply(inputs_, block.graph)
+            expr.debug.ast = node
+        elif isinstance(node, ast.Name):
+            expr = block.read(node.id)
+        elif isinstance(node, ast.Num):
+            expr = self.environment.map(node.n)
+        return expr
+
+    def process_statements(self, block: 'Block',
+                           nodes: List[ast.stmt]) -> 'Block':
+        """Process a sequence of statements."""
+        for node in nodes:
+            block = self.process_statement(block, node)
+        return block
+
+    def process_statement(self, block: 'Block', node: ast.stmt) -> 'Block':
+        """Process a single statement."""
+        if isinstance(node, ast.Assign):
+            return self.process_assign(block, node)
+        elif isinstance(node, ast.FunctionDef):
+            return self.process_function(block, node)
+        elif isinstance(node, ast.Return):
+            return self.process_return(block, node)
+        elif isinstance(node, ast.If):
+            return self.process_if(block, node)
+        elif isinstance(node, ast.While):
+            return self.process_while(block, node)
+        else:
+            raise NotImplementedError(node)
+
+    def process_if(self, block: 'Block', node: ast.If) -> 'Block':
+        """Process a conditional statement.
+
+        A conditional statement generates 3 functions: The true branch, the
+        false branch, and the continuation.
+
+        """
+        true_block = Block(self)
+        false_block = Block(self)
+        after_block = Block(self)
```

Comment:
You have to add the current block to `true/false_block.preds` here, or before the statements are processed in both branches. That seemed enough to make `if` work on my local copy, don't know if there's anything more to do.